README.md

<p align="center">
    <a href="https://github.com/Mrminfive/wxapp-canvas" rel="noopener noreferrer"><img src="./logo.png" alt="Wxapp Canvas logo"></a>
</p>
<p align="center">
    <a href="https://npmcharts.com/compare/wxapp-canvas?minimal=true"><img src="https://img.shields.io/npm/dm/wxapp-canvas.svg" alt="Downloads"></a>
    <a href="https://www.npmjs.com/package/wxapp-canvas"><img src="https://img.shields.io/npm/v/wxapp-canvas.svg" alt="Version"></a>
    <a href="https://www.npmjs.com/package/wxapp-canvas"><img src="https://img.shields.io/npm/l/wxapp-canvas.svg" alt="License"></a>
</p>

<h3 align="center">wxapp-canvas</h3>

<h3 align="center">🎨 微信小程序快速绘画工具</h3>

## 简介

项目使用样式表来绘制 canvas，利用 wxss 下 rpx 单位实现适配，既能保证开发效率，同时也没有学习成本。

组件实现了 css 基本盒子模型，涵盖宽高、背景、字体、边框、内边距等，支持 **自适应宽高**、**文本换行**、**圆角边框**、**背景渐变**、**背景重复** 等好玩功能，感兴趣的可以了解下。

绘制代码没有与组件抽离，有兴趣的同学可以试试抽离去适配多终端玩玩。

<img src="./demo.gif" style="display: block; margin: auto;" />

> 注意：项目使用 `canvasContext.measureText`、`nodesRef.fields[computedStyle]` 等，请保持基础库 >= 2.1.0

## 快速开始

### 使用之前

在开始使用 wxapp-canvas 之前，请阅读 [微信小程序自定义组件][wx-doc] 的相关文档。

### 如何使用

以原生小程序为例：

1. 安装 npm 依赖。

    ``` shell
    npm install --save wxapp-canvas
    ```

    或着将依赖下的 `dist` 目录下的文件拷贝到自己的项目中，建议放置在 `packages/wxapp-canvas` 目录下，组件路径根据自己项目位置配置。

    ``` shell
    ├─packages
    │  └─wxapp-canvas
    │      ├─babel-runtime
    │      │  └─helpers
    │      ├─behaviors
    │      │  └─element
    │      └─components
    │          ├─wxapp-canvas
    │          └─wxapp-canvas-div
    └─pages
        └─index
    ```

2. 添加页面的 json 配置（手动拷贝文件时请注意引用路径）

    ``` json
    "usingComponents": {
        "wxapp-canvas": "wxapp-canvas/components/wxapp-canvas/index",
        "wxapp-canvas-div": "wxapp-canvas/components/wxapp-canvas-div/index"
    }
    ```


3. 在 wxml 中使用组件

    ``` html
    <wxapp-canvas id="wxapp-canvas" class-name="m-wxapp-canvas">
        <wxapp-canvas-div class-name="m-wxapp-canvas-div" csstext="content: '这是节点的内容'; background-color: red;"></wxapp-canvas-div>
    </wxapp-canvas>
    ```
    
    * wxapp-canvas： 
        * `class-name`: 外部样式类，用来代替 `class`
        * `preload`: 配置是否预加载资源
    * wxapp-canvas-div:
        * `class-name`: 外部样式类，用来代替 `class`
        * `csstext`: 外部样式，用来代替 `style`

4. 在 wxss 中编写相应样式

    ``` css
    .m-wxapp-canvas {
        width: 100vw;
        height: 50vh;
    }

    .m-wxapp-canvas-div {
        position: absolute;
        bottom: 20rpx;
        right: 20rpx;
        width: 50%;
        height: 50%;
        border: 10rpx solid;
        border-top-color: red;
        border-right-color: blue;
        border-bottom-color: green;
        border-left-color: yellow;
    }
    ```

    与标准样式层叠表一致，支持基本盒子模型，包括：

    * dimension: 支持宽高自适应、最大最小宽高
    * border：支持基本边框、圆角、阴影
    * background：支持背景颜色、图片、线性渐变、repeat、大小
    * color：兼容标准样式表
    * font：兼容标准样式表
    * padding：兼容标准样式表
    * positioning：支持 top、left、bottom、right、z-index
    * text: 支持 text-align、vertical-align 用于节点内容位置的调节

    与标准和模型不一致的有：

    * position 仅支持 `absolute`
    * 不支持 display
    * 不支持嵌套节点
    * 在有阴影及边框的情况，背景将无法实现透明
    * 渐变背景请以 `linear-gradient(<角度>, [<颜色> <位置>])` 的标准方式书写，如：`linear-gradient(30deg, #fff 0%, #000 100%)`

5. 在 js 中执行渲染

    ``` js
    Page({
        myMethod() {
            wx.showLoading({ title: '绘制中' });
            const canvas = this.selectComponent('#wxapp-canvas');
            canvas
                .draw()
                .then(() => {
                    wx.hideLoading();
                    // 执行自定义操作，如保存为图片

                    canvas.getContext().anvasToTempFilePath(...)
                });
        }
    });
    ```

    组件对外暴露以下方法：

    * `draw`：绘制面板
         * @return {Promise}
    * `getContext`：导出画板上下文
         * @return {Object} canvas 上下文
    * `adaptationText`：适配文本
         * @param {String} str 需要适配的文本
         * @param {Object} font 文本样式
         * @param {Number} maxWidth 最大容器宽度
         * @return {Array} 分段好的文本

## 自定义扩展

wxapp-canvas 允许用户扩展自定义渲染节点。

### 一个栗子

一个基本的自定义渲染节点结构如下：

1. wxml 内容
    ``` html
    <!-- 自定义节点需要设置一个空 view -->
    <view id="canvas-element" class="class-name" style="display: none;"></view>
    ```

2. js 脚本内容
    ``` js
    import { behaviors } from 'wxapp-canvas/index.js';

    Component({
        externalClasses: ['class-name'],

        behaviors: [
            behaviors.element
        ],

        relations: {
            'element': {
                type: 'ancestor',
                target: behaviors.wxappCanvasBehavior,
                linked(target) {
                    this.$canvas = target;
                }
            }
        },

        methods: {
            /**
             * 用于自定义渲染内容
             */
            render(ctx) {
                // do some thing
            },

            /**
             * 用于预加载资源
             *
             * @return {Promise}
             */
            preload(rect, utils) {
                return this._preload(rect, utils)
                    .then(() => {
                        // do some thing
                    });
            }
        }
    });
    ```

### 栗子说明

组件提供 `behaviors.element` 和 `behaviors.wxappCanvasBehavior` 两个 behavior，一个代表节点一个代表容器，任何自定义节点均需像上文例子一样集成 behaviors。

自定义组件还需配置外部样式类，以保证对外规格的统一。

自定义组件有两个必需的方法：`render` 和 `preload`：

* `render`：接收 canvas 上下文，执行自定义绘画动作
* `preload`: 资源预加载，接收容器 `boundingClientRect` 及适配工具脚本，该方法必须调用 `_preload` 解析样式方法及返回 Promise 函数。

## 预览组件

clone 项目，安装以来，用微信开发者工具打开项目下的 example（注意替换项目的 appid）：

``` shell
# clone 项目
git clone https://github.com/Mrminfive/wxapp-canvas.git

# 安装依赖
cd wxapp-canvas && npm install

# 编译组件
npm run dev
```

## License

[MIT](./LICENSE)

Copyright (c) 2018 minfive

[wx-doc]: https://developers.weixin.qq.com/miniprogram/dev/framework/custom-component/


CHANGELOG.md
## CHANGELOG

### v1.0.1 2018-09-08

#### Fix

* 修正构建脚手架
* 移除 dist 文件
* 增加对 miniapp_npm 的支持

------

### v1.0.0 2018-07-15

#### New

* 发布正式版本 v1.0.0

------

### v0.1.0 2018-06-24

#### New

* 初始化项目

------

src/behaviors/element/format-style.js

/**
 * @description 样式格式画函数集
 * @author: minfive
 * @createDate: 2018-06-24
 * @lastModify minfive
 * @lastDate: 2018-06-24
 */

import { errorInfo } from '../../utils.js';

const splitReg = /\s(?!([^()]*\))(?!\)))/;
const systemInfo = wx.getSystemInfoSync();

/**
 * 解析像素值
 *
 * @param {String} setting 设置的值
 * @param {Number} total 100%的实际像素值
 * @param {Number} max 最大值
 * @return {Number} 实际像素值
 */
function processPixels(setting, total, max) {
    let result;

    if (setting === 'auto') {
        result = null;
    } else if (/^calc\(.+\)$/.test(setting)) {
        let match = setting.match(/^calc\((.+)\s\+\s(.+)\)$/);
        result = processPixels(match[1], total) + processPixels(match[2], total);
    } else if (/^.*%$/.test(setting)) {
        result = total * (setting.substring(0, setting.length - 1) / 100);
    } else {
        result = parseFloat(setting);
    }

    return result != null && max != null
        ? result > max ? max : result
        : result;
}

/**
 * 格式化字体样式
 *
 * @param {Object} style 样式配置
 * @return {Object} 字体配置
 */
function formatFontStyle(style) {
    const size = Math.round(processPixels(style['font-size'], 20));
    return {
        color: style['color'],
        style: style['font-style'],
        variant: style['font-variant'],
        weight: style['font-weight'],
        stretch: style['font-stretch'],
        size,
        lineHeight: style['line-height'] === 'normal'
            ? 1.33 * size
            : parseFloat(style['line-height']),
        family: style['font-family'],
        textAlign: ['center', 'left', 'right'].includes(style['text-align']) ? style['text-align'] : 'left',
        verticalAlign: ['top', 'middle', 'bottom'].includes(style['vertical-align']) ? style['vertical-align'] : 'top'
    };
}

/**
 * 格式化盒模型大小
 *
 * @param {Object} computedStyle 未处理的样式配置
 * @param {Object} style 样式配置
 * @param {Object} rect 容器规格
 * @param {Object} utils 工具方法
 * @param {Function} utils.adaptationText 文本内容适配器
 * @param {Function} utils.measureText 文本测量
 * @return {Object} 规格
 */
function formatSize(computedStyle, style, rect, utils) {
    const { adaptationText, measureText } = utils;
    const font = `${style.font.style} ${style.font.weight} ${style.font.size}px ${style.font.family}`;
    const contentText = systemInfo.platform !== 'ios' ? computedStyle.content.slice(1, -1) : computedStyle.content;
    const { border, padding } = style;
    const result = {
        width: processPixels(computedStyle['width'], rect.width),
        height: processPixels(computedStyle['height'], rect.height),
        minWidth: computedStyle['min-width'] === 'none' ? null : processPixels(computedStyle['min-width'], rect.width),
        maxWidth: computedStyle['max-width'] === 'none' ? null : processPixels(computedStyle['max-width'], rect.width),
        minHeight: computedStyle['min-height'] === 'none' ? null : processPixels(computedStyle['min-height'], rect.height),
        maxHeight: computedStyle['max-height'] === 'none' ? null : processPixels(computedStyle['max-height'], rect.height),
        containerSize: {
            width: null,
            height: null
        }
    };

    if (result.minWidth != null && result.maxWidth != null && result.minWidth > result.maxWidth) {
        result.maxWidth = result.minWidth;
    }

    if (result.maxHeight != null && result.maxHeight != null && result.minHeight > result.maxHeight) {
        result.maxHeight = result.minHeight;
    }

    if (result.width == null) {
        result.width = measureText(contentText, font) + border.left.width + border.right.width + padding.left + padding.right;
    }
    result.minWidth && result.width < result.minWidth && (result.width = result.minWidth);
    result.maxWidth && result.width > result.maxWidth && (result.width = result.maxWidth);
    result.containerSize.width = result.width - border.left.width - border.right.width - padding.left - padding.right;
    result.content = adaptationText(contentText, font, result.containerSize.width);

    if (result.height == null) {
        result.height = (border.top.width + border.bottom.width) + (padding.top + padding.bottom) + style.font.lineHeight * result.content.length;
    }
    result.minHeight && result.height < result.minHeight && (result.height = result.minHeight);
    result.maxHeight && result.height > result.maxHeight && (result.height = result.maxHeight);
    result.containerSize.height = result.height - border.top.width - border.bottom.width - padding.top - padding.bottom;

    return result;
}

/**
 * 格式化背景样式
 *
 * @param {Object} style 背景配置
 * @param {Object} rect 容器规格
 * @param {Number} rect.width 容器宽
 * @param {Number} rect.height 容器高
 * @return {Object} 背景配置
 */
function formatBackground(style, rect) {
    function filterSize(size) {
        let defaultVals = ['auto', 'cover', 'contain'];
        let match = size.split(splitReg).filter(item => item != null);

        if (match.length === 2) {
            return [
                processPixels(match[0], rect.width),
                processPixels(match[1], rect.height)
            ];
        } else if (match.length === 1 && !defaultVals.includes(match[0])) {
            return [
                processPixels(match[0], rect.width),
                processPixels(match[0], rect.height)
            ];
        } else {
            return size;
        }
    }

    let setting = {
        color: style['background-color'],
        origin: style['background-origin'],
        clip: style['background-clip'],
        position: {
            // TODE: 图像定位
            // x: rect.width / 2 - processPixels(style['background-position'].split(' ')[0], rect.width),
            // y: rect.height / 2 - processPixels(style['background-position'].split(' ')[1], rect.height)
            x: 0, y: 0
        },
        repeat: style['background-repeat'],
        size: filterSize(style['background-size']),
        image: style['background-image']
    };
    // linear-gradient(50deg, rgb(0, 0, 0), rgb(255, 0, 0) 50%, rgb(0, 153, 0))
    if (setting.image.match(/^url\("*([^"]+)"*\)$/)) {
        setting._imageType = 'image';
        setting._imageUrl = RegExp.$1;
    } else if (setting.image.match(/^linear-gradient\((.+)\)$/)) {
        let match = RegExp.$1.split(/,\sr/g).map((item, idx) => idx === 0 ? item : 'r' + item);
        let angle = ~match[0].indexOf('rgb') ? 180 : parseFloat(match.shift());

        setting._imageType = 'linear-gradient';
        setting._imageGradient = {
            angle,
            stopPoint: match.map(item => {
                let ms = item.match(/^(rgb\(.+\))\s*(\S*)$/);
                if (ms[2] === '') errorInfo('The linear gradient format should follow "linear-gradient(50deg, #000 0%, #f00 50%, #090 100%)"');

                return [ms[1], ms[2]];
            })
        };
    }
    return setting;
}

/**
 * 格式化边框样式
 *
 * @param {Object} border 各个边框的配置
 * @param {String} border.top 顶部边框
 * @param {String} border.left 左部边框
 * @param {String} border.right 右部边框
 * @param {String} border.bottom 底部边框
 * @return {Object} 各个边框的配置
 */
function formatBorderStyle(border) {
    return Object
        .keys(border)
        .reduce((result, key) => {
            let match = border[key].split(splitReg).filter(item => item != null);

            result[key] = {
                width: parseFloat(match[0]),
                style: match[1],
                color: match[2]
            };

            return result;
        }, {});
}

/**
 * 格式化borderRaidus
 *
 * @param {String} csstext 边角样式
 * @param {Object} rect 容器规格
 * @param {Number} rect.width 容器宽
 * @param {Number} rect.height 容器高
 * @return {Array} 边角在水平及垂直方向上的半径
 */
function formatBorderRadiusStyle(csstext, rect) {
    let match = csstext.split(' ');

    if (match.length === 1) {
        match = Array(2).fill(match[0]);
    }

    return [
        processPixels(match[0], rect.width, rect.width / 2),
        processPixels(match[1], rect.height, rect.height / 2)
    ];
}

/**
 * 格式化 boxShadow
 *
 * @param {String} csstext 阴影样式
 * @return {Array} 阴影配置
 */
function formatBoxShadow(csstext) {
    let shadows = csstext === 'none'
        ? []
        : csstext.split(', r')
            .filter(set => set != null)
            .map((set, idx) => {
                let match = (idx === 0 ? set : 'r' + set)
                    .split(splitReg)
                    .filter(item => item != null);
                return {
                    color: match[0],
                    offsetX: parseFloat(match[1]),
                    offsetY: parseFloat(match[2]),
                    blur: parseFloat(match[3])
                };
            });

    return shadows;
}

/**
 * 格式化样式配置
 *
 * @param {Object} computedStyle 获取到的样式配置
 * @param {Object} rect canvas boundingClientRect
 * @param {Object} utils 工具方法
 * @param {Function} utils.adaptationText 文本内容适配器
 * @param {Function} utils.measureText 文本测量
 * @return {Object}
 */
function formatStyle(computedStyle, rect, utils) {
    let style = {
        position: 'absolute',
        left: processPixels(computedStyle.left, rect.width),
        top: processPixels(computedStyle.top, rect.height),
        right: processPixels(computedStyle.right, rect.width),
        bottom: processPixels(computedStyle.bottom, rect.width),
        zIndex: computedStyle['z-index'] === 'auto' ? 0 : computedStyle['z-index'],
        overflow: ['hidden', 'visible'].includes(computedStyle['overflow']) ? computedStyle['overflow'] : 'visible',

        display: computedStyle.display,
        padding: {
            top: parseFloat(computedStyle['padding-top']),
            left: parseFloat(computedStyle['padding-left']),
            right: parseFloat(computedStyle['padding-right']),
            bottom: parseFloat(computedStyle['padding-bottom'])
        },
        boxShadow: formatBoxShadow(computedStyle['box-shadow']),

        font: formatFontStyle(computedStyle),

        border: formatBorderStyle({
            top: computedStyle['border-top'],
            left: computedStyle['border-left'],
            right: computedStyle['border-right'],
            bottom: computedStyle['border-bottom']
        })
    };

    Object.assign(style, formatSize(computedStyle, style, rect, utils));

    style.left === null && style.right === null && (style.left = 0);
    style.top === null && style.bottom === null && (style.top = 0);

    let { padding, width, height } = style;
    return {
        ...style,
        background: formatBackground(
            computedStyle,
            {
                width: style.width - padding.left - padding.right,
                height: style.height - padding.top - padding.bottom
            }
        ),
        borderRadius: {
            'top-left': formatBorderRadiusStyle(computedStyle['border-top-left-radius'], { width, height }),
            'top-right': formatBorderRadiusStyle(computedStyle['border-top-right-radius'], { width, height }),
            'bottom-left': formatBorderRadiusStyle(computedStyle['border-bottom-left-radius'], { width, height }),
            'bottom-right': formatBorderRadiusStyle(computedStyle['border-bottom-right-radius'], { width, height })
        },
        startX: style.left === null
            ? rect.width - style.right - width
            : style.right === null
                ? style.left
                : style.left >= style.right
                    ? style.left
                    : rect.width - style.right - width,
        startY: style.top === null
            ? rect.height - style.bottom - height
            : style.bottom === null
                ? style.top
                : style.top >= style.bottom
                    ? style.top
                    : rect.height - style.bottom - height
    };
}

export {
    processPixels,
    formatSize,
    formatFontStyle,
    formatBackground,
    formatBorderStyle,
    formatBorderRadiusStyle,
    formatBoxShadow,
    formatStyle
};
